Source/WebCore/css/CSSBasicShapes.cpp

```diff
@@ -191,6 +191,116 @@ bool CSSEllipseValue::equals(const CSSEllipseValue& other) const
+}
+
+static inline void updateCornerRadiusWidthAndHeight(const CSSValue* corner, String& width, String& height)
+{
+    if (!corner)
+        return;
+    width = corner->first().cssText();
+    height = corner->second().cssText();
+}
+
+static bool buildRadii(Vector<String>& radii, const String& topLeftRadius, const String& topRightRadius, const String& bottomRightRadius, const String& bottomLeftRadius)
+{
+    bool showBottomLeft = topRightRadius != bottomLeftRadius;
+    bool showBottomRight = showBottomLeft || (bottomRightRadius != topLeftRadius);
+    bool showTopRight = showBottomRight || (topRightRadius != topLeftRadius);
+
+    radii.append(topLeftRadius);
+    if (showTopRight)
+        radii.append(topRightRadius);
+    if (showBottomRight)
+        radii.append(bottomRightRadius);
+    if (showBottomLeft)
+        radii.append(bottomLeftRadius);
+
+    return radii.size() == 1 && radii[0] == "0px"_s;
+}
+
+static String buildXywhString(const String& insetX, const String& insetY, const String& width, const String& height,
+    const String& topLeftRadiusWidth, const String& topLeftRadiusHeight,
+    const String& topRightRadiusWidth, const String& topRightRadiusHeight,
+    const String& bottomRightRadiusWidth, const String& bottomRightRadiusHeight,
+    const String& bottomLeftRadiusWidth, const String& bottomLeftRadiusHeight)
+{
+    StringBuilder result;
+    result.append("xywh("_s, insetX, ' ', insetY, ' ', width, ' ', height);
+
+    if (!topLeftRadiusWidth.isNull() && !topLeftRadiusHeight.isNull()) {
+        Vector<String> horizontalRadii;
+        bool areDefaultCornerRadii = buildRadii(horizontalRadii, topLeftRadiusWidth, topRightRadiusWidth, bottomRightRadiusWidth, bottomLeftRadiusWidth);
+
+        Vector<String> verticalRadii;
+        areDefaultCornerRadii &= buildRadii(verticalRadii, topLeftRadiusHeight, topRightRadiusHeight, bottomRightRadiusHeight, bottomLeftRadiusHeight);
+
+        if (!areDefaultCornerRadii) {
+            result.append(" round"_s);
+
+            for (auto& radius : horizontalRadii)
+                result.append(' ', radius);
+
+            if (verticalRadii.size() != horizontalRadii.size()
+                || !WTF::VectorComparer<false, String>::compare(verticalRadii.data(), horizontalRadii.data(), verticalRadii.size())) {
+                result.append(" /"_s);
+                for (auto& radius : verticalRadii)
+                    result.append(' ', radius);
+            }
+        }
+    }
+    result.append(')');
+    return result.toString();
+}
+
+String CSSXywhValue::customCSSText() const
+{
+    String topLeftRadiusWidth;
+    String topLeftRadiusHeight;
+    String topRightRadiusWidth;
+    String topRightRadiusHeight;
+    String bottomRightRadiusWidth;
+    String bottomRightRadiusHeight;
+    String bottomLeftRadiusWidth;
+    String bottomLeftRadiusHeight;
+
+    updateCornerRadiusWidthAndHeight(topLeftRadius(), topLeftRadiusWidth, topLeftRadiusHeight);
```

Comment:
It would be much more efficient to find a way to build one string without intermediate strings.

Reply:
This code shared with inset(), so cleaning up would be better as a followup.